doc/catalyst-cli/catalyst-cli.rst

```diff
@@ -0,0 +1,215 @@
+* ``translate``: Run ``mlir-translate`` on the LLVM MLIR dialect input.
+* ``llc``: Run ``llc`` on the LLVM IR input.
+* ``all``: Run all of ``opt``, ``translate`` and ``llc`` on the MLIR input.
+
+``--save-ir-after-each=<pass|pipeline>``
+""""""""""""""""""""""""""""""""""""""""
+
+Keep intermediate files after each pass or after each pipeline in the compilation. By default, no
+intermediate files are saved.
+
+* ``pass``: Keep intermediate files after each transformation/optimization pass.
+* ``pipeline``: Keep intermediate files after each pipeline, where a *pipeline* is a sequence of
+  transformation/optimization passes.
+
+``--keep-intermediate[=<true|false>]``
+""""""""""""""""""""""""""""""""""""""
+
+Keep intermediate files after each pipeline in the compilation. By default, no intermediate files
+are saved. Using ``--keep-intermediate`` is equivalent to using ``--save-ir-after-each=pipeline``.
+
+``--catalyst-pipeline=<pipeline1(pass1[;pass2[;...]])[,pipeline2(...)]>``
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Specify the Catalyst compilation pass pipelines.
+
+A pipeline is composed of a semicolon-delimited sequence of one or more transformation or
+optimization passes. Multiple pass pipelines can be specified and input as a comma-delimited
+sequence of pipelines.
+
+For example, if we wanted to specify two pass pipelines, ``pipe1`` and ``pipe2``, where ``pipe1``
+applies the passes ``split-multiple-tapes`` and ``apply-transform-sequence``, and where ``pipe2``
+applies the pass ``inline-nested-module``, we would specifiy this pipeline configuration as:
+
+.. code-block::
+
+    --catalyst-pipeline=pipe1(split-multiple-tapes;apply-transform-sequence),pipe2(inline-nested-module)
+
+``--workspace=<path>``
+""""""""""""""""""""""
+
+The workspace directory where intermediate files are saved, and from which they are read when using
+the ``--checkpoint-stage`` option. The default is the current working directory. Note that the
+workspace directory must exist before running ``catalyst-cli`` with this option.
+
+``--module-name=<name>``
+""""""""""""""""""""""""
+
+The module name used in naming the output file(s). The default is ``"catalyst_module"``. Using the
+``-o`` option to specify the output filename overrides this option.
+
+``--async-qnodes[=<true|false>]``
+"""""""""""""""""""""""""""""""""
+
+Enable asynchronous QNodes.
+
+``--checkpoint-stage=<stage name>``
+"""""""""""""""""""""""""""""""""""
+
+Define a *checkpoint stage*, used to indicate that the compiler should start only after reaching the
+given pass.
+
+``--dump-catalyst-pipeline[=<true|false>]``
+"""""""""""""""""""""""""""""""""""""""""""
+
+Print (to stderr) the pipeline(s) that will be run.
+
+Examples
+^^^^^^^^
+
+To illustrate how to use the Catalyst CLI tool, consider the very simple MLIR code, ``foo.mlir``,
+which defines a function ``foo`` that takes in no arguments and returns nothing:
+
+.. code-block::
```

Comment:
In this example, all three passes do nothing. The fact that the passes put in an extra enclosing module is merely an optical illusion: there is always a implied top-level module in an mlir file, even if none is specified explicitly in the file itself. When dumping IR, this implicit top-level module becomes explicitly dumped.

As a user reading the documentation, I would feel more comfortable if I knew what the passes did. I suggest we put in the implicit module in the input mlir, and say in the documentation that the three passes do nothing on this simple example. What do you think?

Reply:
That sounds reasonable to me. It might be even nicer if we can come up with an example that can actually exercise at least one of the passes, to show that something is actually happening! Although even the simplest examples require many lines of MLIR, and for the sake of documentation it would probably be better to keep the examples as concise as possible.

I'll add the top-level module for now, maybe we can come up with a better example later.

Reply:
How about a peephole example? mlir does not have to be a fully well-produced one from the frontend, e.g. (adopted from [a mlir test](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/test/Quantum/ChainedSelfInverseTest.mlir#L34)
```
func.func @my_circuit(%in_qubit: !quantum.bit, %angle: f64) -> !quantum.bit {
    %0 = quantum.custom "RX"(%angle) %in_qubit : !quantum.bit
    %1 = quantum.custom "Hadamard"() %0 : !quantum.bit
    %2 = quantum.custom "Hadamard"() %1 : !quantum.bit
    %3 = quantum.custom "RX"(%angle) %2 : !quantum.bit
    return %3 : !quantum.bit
}
```
and send this example through the `cancel inverse, merge rotation` pipeline?

As a bonus, we can show that the reverse pipeline `merge rotation, cancel inverse` produces a different result, if there is enough paper space.

Reply:
Added in e550709e

Reply:
```suggestion
.. code-block:: mlir
```